File: backend/utils/cpe_utils.py
"""
Utility functions for parsing CPE (Common Platform Enumeration) data.
"""

from typing import List, Dict, Any

def extract_affected_products(configurations: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
    """
    Extract affected products from vulnerability configurations.
    
    Args:
        configurations: List of vulnerability configuration objects
        
    Returns:
        List of affected product dictionaries
    """
    products = []
    seen = set()
    
    for config in configurations:
        nodes = config.get("nodes", [])
        for node in nodes:
            cpe_matches = node.get("cpeMatch", [])
            for cpe in cpe_matches:
                if cpe.get("vulnerable", False):
                    criteria = cpe.get("criteria", "")
                    if criteria.startswith("cpe:"):
                        parsed = parse_cpe(criteria)
                        if parsed:
                            # Build version range if available
                            version_range = parsed.get("version")
                            if (cpe.get("versionStartIncluding") or cpe.get("versionEndIncluding") or 
                                cpe.get("versionStartExcluding") or cpe.get("versionEndExcluding")):
                                ranges = []
                                if cpe.get("versionStartIncluding"):
                                    ranges.append(f">= {cpe['versionStartIncluding']}")
                                if cpe.get("versionStartExcluding"):
                                    ranges.append(f"> {cpe['versionStartExcluding']}")
                                if cpe.get("versionEndIncluding"):
                                    ranges.append(f"<= {cpe['versionEndIncluding']}")
                                if cpe.get("versionEndExcluding"):
                                    ranges.append(f"< {cpe['versionEndExcluding']}")
                                version_range = ", ".join(ranges)
                            
                            product_key = f"{parsed['vendor']}:{parsed['product']}:{version_range or 'any'}"
                            if product_key not in seen:
                                seen.add(product_key)
                                product_info = {
                                    "vendor": parsed["vendor"],
                                    "product": parsed["product"],
                                    "version": version_range
                                }
                                products.append(product_info)
    
    return products

def parse_cpe(cpe_string: str) -> Dict[str, str]:
    """
    Parse CPE criteria string to extract vendor, product, and version information.
    
    CPE format: cpe:2.3:part:vendor:product:version:update:edition:language:sw_edition:target_sw:target_hw:other
    
    Args:
        cpe_string: CPE criteria string
        
    Returns:
        Dictionary with vendor, product, and version information
    """
    try:
        if not cpe_string or not isinstance(cpe_string, str):
            return None
        
        parts = cpe_string.split(":")
        if len(parts) < 6 or parts[0] != "cpe":
            return None
        
        vendor = parts[3] if parts[3] not in ("*", "-") else "Unknown"
        product = parts[4] if parts[4] not in ("*", "-") else "Unknown"
        version = parts[5] if parts[5] not in ("*", "-") else None
        
        # Clean up underscores and decode URI components
        vendor = vendor.replace("_", " ").title()
        product = product.replace("_", " ").title()
        if version:
            version = version.replace("_", " ")
        
        return {
            "vendor": vendor,
            "product": product,
            "version": version
        }
        
    except Exception:
        return None

File: frontend/src/utils/cpeUtils.ts
// Utility functions for parsing CPE (Common Platform Enumeration) data

export interface AffectedProduct {
    vendor: string;
    product: string;
    version?: string;
    versionRange?: string;
}

/**
 * Parse CPE criteria string to extract vendor, product, and version information
 * CPE format: cpe:2.3:part:vendor:product:version:update:edition:language:sw_edition:target_sw:target_hw:other
 */
export function parseCPE(cpeString: string): AffectedProduct | null {
    try {
        if (!cpeString || typeof cpeString !== 'string') {
            return null;
        }

        const parts = cpeString.split(':');

        if (parts.length < 6 || parts[0] !== 'cpe') {
            return null;
        }

        const vendor = parts[3] === '*' || parts[3] === '-' ? 'Unknown' : decodeURIComponent(parts[3]).replace(/_/g, ' ');
        const product = parts[4] === '*' || parts[4] === '-' ? 'Unknown' : decodeURIComponent(parts[4]).replace(/_/g, ' ');
        const version = (parts[5] === '*' || parts[5] === '-') ? undefined : decodeURIComponent(parts[5]).replace(/_/g, ' ');

        return {
            vendor: capitalizeWords(vendor),
            product: capitalizeWords(product),
            version: version ? capitalizeWords(version) : undefined
        };
    } catch (error) {
        console.warn('Failed to parse CPE:', cpeString, error);
        return null;
    }
}

/**
 * Extract affected products from vulnerability configurations
 */
export function extractAffectedProducts(configurations: any[]): AffectedProduct[] {
    const products: AffectedProduct[] = [];
    const seen = new Set<string>();

    if (!configurations || configurations.length === 0) {
        return products;
    }

    configurations.forEach((config) => {
        if (!config.nodes || !Array.isArray(config.nodes)) {
            return;
        }

        config.nodes.forEach((node: any) => {
            if (!node.cpeMatch || !Array.isArray(node.cpeMatch)) {
                return;
            }

            node.cpeMatch.forEach((cpe: any) => {
                if (cpe.vulnerable) {
                    const parsed = parseCPE(cpe.criteria);

                    if (parsed) {
                        // Build version range if available
                        let versionRange = parsed.version;
                        if (cpe.versionStartIncluding || cpe.versionEndIncluding ||
                            cpe.versionStartExcluding || cpe.versionEndExcluding) {
                            const ranges = [];
                            if (cpe.versionStartIncluding) ranges.push(`>= ${cpe.versionStartIncluding}`);
                            if (cpe.versionStartExcluding) ranges.push(`> ${cpe.versionStartExcluding}`);
                            if (cpe.versionEndIncluding) ranges.push(`<= ${cpe.versionEndIncluding}`);
                            if (cpe.versionEndExcluding) ranges.push(`< ${cpe.versionEndExcluding}`);
                            versionRange = ranges.join(', ');
                        }

                        const productKey = `${parsed.vendor}:${parsed.product}:${versionRange || 'any'}`;
                        if (!seen.has(productKey)) {
                            seen.add(productKey);
                            products.push({
                                ...parsed,
                                versionRange
                            });
                        }
                    }
                }
            });
        });
    });

    return products;
}

/**
 * Get a summary of affected products for display in tables
 */
export function getAffectedProductsSummary(configurations: any[]): string {
    const products = extractAffectedProducts(configurations);

    if (products.length === 0) {
        return 'No specific products';
    }

    if (products.length === 1) {
        const product = products[0];
        return `${product.vendor} ${product.product}`;
    }

    if (products.length <= 3) {
        return products.map(p => `${p.vendor} ${p.product}`).join(', ');
    }

    const first = products[0];
    return `${first.vendor} ${first.product} and ${products.length - 1} others`;
}

/**
 * Capitalize words in a string
 */
function capitalizeWords(str: string): string {
    return str.split(' ')
        .map(word => word.charAt(0).toUpperCase() + word.slice(1).toLowerCase())
        .join(' ');
}